cmd/server.go

```diff
@@ -0,0 +1,352 @@
+func doServer(cmd *cobra.Command, args []string) {
+	listener, err := net.Listen("tcp", otlpEndpoint)
+	if err != nil {
+		log.Fatalf("failed to listen: %s", err)
+	}
+
+	cs := cliServer{stopper: make(chan bool)}
+	if serverConf.pterm {
+		// TODO: implement a limit on this, drop old spans, etc.
+		cs.events = []CliEvent{}
+
+		area, err := pterm.DefaultArea.Start()
+		if err != nil {
+			log.Fatalf("failed to set up terminal for rendering: %s", err)
+		}
+		cs.area = area
+	}
+	gs := grpc.NewServer()
+	v1.RegisterTraceServiceServer(gs, &cs)
+
+	// stops the grpc server after timeout
+	if serverConf.timeout > 0 {
+		go func() {
+			time.Sleep(time.Duration(serverConf.timeout) * time.Second)
+			cs.stopper <- true
+		}()
+	}
+
+	// single place to stop the server, used by timeout and max-spans
+	go func() {
+		<-cs.stopper
+		if serverConf.pterm {
+			cs.area.Stop()
+		}
+		gs.Stop()
+	}()
+
+	if err := gs.Serve(listener); err != nil {
+		log.Fatalf("failed to serve: %s", err)
+	}
+}
+
+// cliServer is a gRPC/OTLP server handle.
+type cliServer struct {
+	spansSeen int
+	stopper   chan bool
+	events    CliEventList // a queue for pterm drawing mode
+	area      *pterm.AreaPrinter
+	v1.UnimplementedTraceServiceServer
+}
+
+// Export implements the gRPC server interface for exporting messages.
+func (cs *cliServer) Export(ctx context.Context, req *v1.ExportTraceServiceRequest) (*v1.ExportTraceServiceResponse, error) {
+	rss := req.GetResourceSpans()
+	for _, resource := range rss {
+		ilSpans := resource.GetInstrumentationLibrarySpans()
+		for _, ils := range ilSpans {
+			for _, span := range ils.GetSpans() {
+				// convert protobuf spans to something easier for humans to consume
+				ces := newCliEventFromSpan(span, ils)
+				events := []CliEvent{}
+				for _, se := range span.GetEvents() {
+					events = append(events, newCliEventFromSpanEvent(se, span, ils))
+				}
+
+				if serverConf.outDir != "" {
+					cs.writeFile(ces, events)
+				}
+
+				if serverConf.pterm {
+					cs.drawPterm(ces, events)
+				}
+
+				cs.spansSeen++ // count spans for exiting on --max-spans
+				if serverConf.maxSpans > 0 && cs.spansSeen >= serverConf.maxSpans {
+					cs.stopper <- true // shus the server down
+					return &v1.ExportTraceServiceResponse{}, nil
+				}
+			}
+		}
+	}
+
+	return &v1.ExportTraceServiceResponse{}, nil
+}
+
+// writeFile takes the spans and events and writes them out to json files in the
+// tid/sid/span.json and tid/sid/events.json files.
+func (cs *cliServer) writeFile(span CliEvent, events []CliEvent) {
+	// create trace directory
+	outpath := filepath.Join(serverConf.outDir, span.TraceID)
+	os.Mkdir(outpath, 0755) // ignore errors for now
+
+	// create span directory
+	outpath = filepath.Join(outpath, span.SpanID)
+	os.Mkdir(outpath, 0755) // ignore errors for now
+
+	// write span to file
+	sjs, _ := json.Marshal(span)
+	spanfile := filepath.Join(outpath, "span.json")
+	err := os.WriteFile(spanfile, sjs, 0644)
```

Comment:
This truncates and re-writes the file every time (I think). Is that what we want?

Reply:
probably? are there cases where a client will send the same span twice? I'll add a TODO in the code

Reply:
I probably misunderstood the control flow here, no worries.